File: minecraft/src/main/java/dev/dediamondpro/minemark/minecraft/elements/MarkdownTextElement.java
/*
 * This file is part of MineMark
 * Copyright (C) 2024 DeDiamondPro
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License Version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.dediamondpro.minemark.minecraft.elements;

import dev.dediamondpro.minemark.LayoutData;
import dev.dediamondpro.minemark.LayoutStyle;
import dev.dediamondpro.minemark.elements.Element;
import dev.dediamondpro.minemark.elements.impl.TextElement;
import dev.dediamondpro.minemark.minecraft.platform.MarkdownRenderer;
import dev.dediamondpro.minemark.minecraft.style.MarkdownStyle;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.xml.sax.Attributes;

import java.awt.*;

public class MarkdownTextElement extends TextElement<MarkdownStyle, MarkdownRenderer> {
    public MarkdownTextElement(@NotNull String text, @NotNull MarkdownStyle style, @NotNull LayoutStyle layoutStyle, @Nullable Element<MarkdownStyle, MarkdownRenderer> parent, @NotNull String qName, @Nullable Attributes attributes) {
        super(text, style, layoutStyle, parent, qName, attributes);
    }

    @Override
    protected void drawText(@NotNull String text, float x, float y, float fontSize, Color color, boolean hovered, LayoutData.MarkDownElementPosition position, @NotNull MarkdownRenderer renderer) {
        renderer.drawText(getPrefix(hovered) + text, (int) x, (int) y, fontSize, color.getRGB(), style.getTextStyle().hasShadow());
    }

    @Override
    protected void drawInlineCodeBlock(float x, float y, float width, float height, Color color, @NotNull MarkdownRenderer renderer) {
        renderer.drawRect((int) x, (int) y, (int) width, (int) height, color.getRGB());
    }

    @Override
    protected float getTextWidth(@NotNull String text, float fontSize, MarkdownRenderer renderer) {
        return renderer.getTextWidth(getPrefix(false) + text, fontSize);
    }

    private String getPrefix(boolean hovered) {
        StringBuilder prefixBuilder = new StringBuilder();
        if (layoutStyle.isBold()) prefixBuilder.append("§l");
        if (layoutStyle.isItalic()) prefixBuilder.append("§o");
        if (layoutStyle.isStrikethrough()) prefixBuilder.append("§m");
        if (layoutStyle.isUnderlined() || layoutStyle.isPartOfLink() && hovered) prefixBuilder.append("§n");
        return prefixBuilder.toString();
    }

    @Override
    protected float getBaselineHeight(float fontSize, MarkdownRenderer renderer) {
        return 8f * fontSize;
    }

    @Override
    protected float getDescender(float fontSize, MarkdownRenderer renderData) {
        return fontSize;
    }
}

File: polyui/src/main/kotlin/dev/dediamondpro/minemark/polyui/elements/MarkdownTextElement.kt
/*
 * This file is part of MineMark
 * Copyright (C) 2024 DeDiamondPro
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License Version 3 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */


package dev.dediamondpro.minemark.polyui.elements

import dev.dediamondpro.minemark.LayoutData.MarkDownElementPosition
import dev.dediamondpro.minemark.LayoutStyle
import dev.dediamondpro.minemark.elements.Element
import dev.dediamondpro.minemark.elements.impl.TextElement
import dev.dediamondpro.minemark.polyui.MarkdownStyle
import org.polyfrost.polyui.color.PolyColor
import org.polyfrost.polyui.color.toPolyColor
import org.polyfrost.polyui.component.Drawable
import org.polyfrost.polyui.component.extensions.events
import org.polyfrost.polyui.event.Event
import org.xml.sax.Attributes
import java.awt.Color
import kotlin.math.floor

class MarkdownTextElement(
    text: String,
    style: MarkdownStyle,
    layoutStyle: LayoutStyle,
    parent: Element<MarkdownStyle, Drawable>?,
    qName: String,
    attributes: Attributes?,
) : TextElement<MarkdownStyle, Drawable>(text, style, layoutStyle, parent, qName, attributes) {
    private val textColor: PolyColor = layoutStyle.textColor.toPolyColor()
    private val codeBlockColor: PolyColor = style.codeBlockStyle.color.toPolyColor()
    private val font =
        when {
            layoutStyle.isPartOfCodeBlock -> style.codeBlockStyle.codeFont
            layoutStyle.isBold && layoutStyle.isItalic -> style.textStyle.italicBoldFont
            layoutStyle.isBold -> style.textStyle.boldFont
            layoutStyle.isItalic -> style.textStyle.italicNormalFont
            else -> style.textStyle.normalFont
        }
    private var lineHeight = -1f
    private var registeredEvent = false
    private var wasHovered = false

    private fun registerEvent(drawable: Drawable) {
        if (registeredEvent) return
        registeredEvent = true

        if (!layoutStyle.isPartOfLink) return
        drawable.events {
            Event.Mouse.Moved.then {
                var hovered = false
                for (position in lines.keys) {
                    if (position.isInside(drawable.polyUI.mouseX, drawable.polyUI.mouseY)) {
                        hovered = true
                        break
                    }
                }
                if (hovered != wasHovered) {
                    wasHovered = hovered
                    drawable.needsRedraw = true
                }
            }
        }
    }

    override fun beforeDrawInternal(
        xOffset: Float,
        yOffset: Float,
        mouseX: Float,
        mouseY: Float,
        drawable: Drawable
    ) {
        registerEvent(drawable)
        super.beforeDrawInternal(xOffset, yOffset, mouseX, mouseY, drawable)
    }

    override fun drawText(
        text: String,
        x: Float,
        y: Float,
        fontSize: Float,
        color: Color,
        hovered: Boolean,
        position: MarkDownElementPosition,
        drawable: Drawable,
    ) {
        if (layoutStyle.isUnderlined || layoutStyle.isPartOfLink && hovered) {
            drawable.renderer.rect(x, position.bottomY, position.width, floor(layoutStyle.fontSize / 8), textColor)
        }
        if (layoutStyle.isStrikethrough) {
            drawable.renderer.rect(
                x,
                y + position.height / 2f - fontSize / 8f,
                position.width,
                floor(layoutStyle.fontSize / 8),
                textColor,
            )
        }
        drawable.renderer.text(font, x, y, text, textColor, fontSize)
    }

    override fun drawInlineCodeBlock(
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        color: Color,
        drawable: Drawable,
    ) {
        drawable.renderer.rect(x, y, width, height, codeBlockColor, radius = 2f)
    }

    override fun getTextWidth(
        text: String,
        fontSize: Float,
        drawable: Drawable,
    ): Float {
        return drawable.renderer.textBounds(font, text, fontSize).x
    }

    override fun getBaselineHeight(
        fontSize: Float,
        drawable: Drawable,
    ): Float {
        if (lineHeight == -1f) {
            lineHeight = drawable.renderer.textBounds(font, text, fontSize).y
        }
        return lineHeight
    }
}
